Fixing the chat client

twilightwanderer opened the pull request:
## About The Pull Request

Corrects the chat restriction on using ASII characters only. ([This PR](https://github.com/Skyrat-SS13/Skyrat-tg/pull/7598) was created for Skyrat-TG, but the collaborator of that repository asked me to adapt this PR for tgstation.)

![изображение](https://user-images.githubusercontent.com/88540658/129605970-cd659a0f-6f12-4a33-b6a7-6e4f36bdb33f.png)

## Why It's Good For The Game

This will fix #54598
In the original code, the chat client uses only ASII standard characters, which is very limited in its capabilities. For example, does not allow you to use specialized characters, which would have taken the atmosphere of the old messengers, as well as regional characters. The lack of regional characters complicates the game for non-English-speaking servers. For example, the Russian-speaking player community Space Station 13 The Fluffy Frontier uses the original Skyrat-tg build. And the players of this community almost do not use the chat client to communicate due to the fact that they can not use Cyrillic characters. 

I tried to solve this problem by rewriting some of the text handling functions, namely:
- /proc/reject_bad_message(text) - сhecks messages for emptiness;
- /proc/reject_bad_username(text, max_length = 32) - Checks the entered name for emptiness and length. (32 characters seems optimal to me for a nickname. But if not, it's easy to change.);
- /proc/reject_bad_chatname(text, max_length = 12) - Checks the entered name of the chat for emptiness and the number of characters (I chose 12 characters because a large number expands the panel with chats to the size in which you can not see the main message box). 

## Changelog
:cl:
add: Adding the ability to use a large number of characters in the chat client.
code: In the chat client code, the standard text checking procedures have been replaced by special chat text checking procedures.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

twilightwanderer committed:
Adding сhecking procedures for the ntnrc_client.dm

Adding procedures to /code/__HELPERS/text.dm to check typed text in chat client ntnrc_client.dm
Added:
- /proc/reject_bad_message(text)
- /proc/reject_bad_username(text, max_length = 32)
- /proc/reject_bad_chatname(text, max_length = 12)

twilightwanderer committed:
Corrections to character limit on ntnrc_client.dm 

The __HELPERS\text.dm procedures, which limit input to characters outside 128 characters in the ASII standard, are disabled. Added procedures for processing and checking input text:
- /datum/computer_file/program/chatclient/proc/reject_bad_message(text)
- /datum/computer_file/program/chatclient/proc/reject_bad_username(text, max_length = 32)
- /datum/computer_file/program/chatclient/proc/reject_bad_chatname(text, max_length = 12)
These procedures allow you to enter regional characters, checking the entered text for emptiness and number of characters.

maxymax13 commented:
other locale support is good but it's not really an irc channel if you can just drop emojis in there

twilightwanderer commented:
> 
> other locale support is good but it's not really an irc channel if you can just drop emojis in there

And what's wrong with using emoji, since almost every player can create his own chat? In any case, it's easy to limit it to a couple of lines in the procedure.

Ghommie commented:
Don't forget to add "this will fix #54598" to the PR description.

twilightwanderer commented:
> Don't forget to add "this will fix #54598" to the PR description.

Added as per your request

twilightwanderer commented:
Should I still hope for some kind of response or is it better to close the PR immediately?

LemonInTheDark commented:
> 
> 
> Should I still hope for some kind of response or is it better to close the PR immediately?

somewhere between 1 day and a month baby.
feel free to bring it up on discord if your pr gets missed, we get a lot and it can be hard to stay on top.

twilightwanderer commented on code/__HELPERS/text.dm:
```diff
@@ -953,3 +953,81 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
+				if(0 to 31)
+					return
+				if(32)
+					continue
```
@Watermelon914  This construction is designed to keep track of invalid characters. [https://en.wikipedia.org/wiki/ASCII#/media/File:USASCII_code_chart.png](ASCII_Code_Chart) Yes, you can put it in a separate procedure, but there is a problem with this solution. The problem is that you have to take out not a part of the code, which you indicated, but the whole loop. Right now this loop looks the same in all three procedures, but if you need to change it at least in one procedure, your solution will immediately lose its meaning.

Watermelon914 replied:
except they there is a lot identical in all 3 functions. This just copy and pasted code. Move the common logic out to its own proc and call them in these procs.

Watermelon914 replied:
In fact, you can get rid of the most code duplication by just calling reject_bad_message after your if statement to check if the length of the text is greater than the max length.

twilightwanderer committed:
Merge branch 'master' into Fixing-the-chat-client

optimumtact commented:
please fix your lint issues.

twilightwanderer commented:
> please fix your lint issues.

I apologize profusely. At the moment I am very busy with my studies and can not find time to finish it. I will try to fix it in the next few days.

twilightwanderer committed:
Unification of the text checking procedure

Combining three text checking procedures in ntnrc_client.dm into one.

twilightwanderer committed:
Replacement of text validation procedures.

Text procedures have been replaced by unitary procedures. Added definitions for text size.

Watermelon914 commented:
your CI is failing

twilightwanderer committed:
Delete space indentation

twilightwanderer committed:
Delete space indentation

LemonInTheDark commented:
Why are these all webedits?

twilightwanderer commented:
> Why are these all webedits?

Because I didn't see them working.

![изображение](https://user-images.githubusercontent.com/88540658/134794512-065b1657-32b4-4445-bb24-80107a67b1be.png)

LemonInTheDark commented:
I'm sorry I don't quite understand

twilightwanderer commented:
> I'm sorry I don't quite understand

Oh, I'm sorry. For some reason I thought of html tags. It's just that in the case of tgstation, I find it more convenient to edit the changes through the github editor. If this is critical, I can send changes from my local repository.

LemonInTheDark commented:
We prefer webedits, especially for changes outside of a few lines. They provide at least a small hope that the code was tested locally, which outside of our qa team (the playerbase) yelling at us is really our only method of quality control.
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process

twilightwanderer commented:
> We prefer standard commits, and would typically require them for changes outside of a few lines. They provide at least a small hope that the code was tested locally, which outside of our qa team (the playerbase) yelling at us is really our only method of quality control outside of the inspection maintainers do
> https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process

I'll keep that in mind. To confirm that I tested my changes, I will record an animation of the chat.

LemonInTheDark commented:
It's fine, I trust you. Just keep it in mind for future yeah?

twilightwanderer commented:
> It's fine, I trust you. Just keep it in mind for future yeah?

Absolutely!

twilightwanderer commented:
> Finally, trollface ascii in chat client

I'm afraid it won't work. To use ASII graphics, you have to redo the text output. I didn't do that.

Watermelon914 commented:
tested the code?

twilightwanderer commented:
> tested the code?

Of course

Watermelon914 commented on code/__HELPERS/text.dm:
```diff
@@ -989,3 +989,27 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
+	if (length(text) > max_length)
+		return
+	else
+		for(var/i = 1, i <= length(text), i += length(char))
```
use `for(var/i in 1 to length(text)`
Alternatively, texts are usually stored as arrays so normally you'd be able to do this too but I'm not 100% sure you can do this in byond, so see if it compiles or not
`for(var/i in text)`

Also, try naming the i variable something more detailed, like index.

LemonInTheDark replied:
I think that'll break for unicode chars, at least if you want the char and not each byte. Same with length() https://secure.byond.com/docs/ref/#/{notes}/Unicode

Watermelon914 replied:
It can at least be changed to 
`for(var/i in 1 to length(text)` as a micro optimization

LemonInTheDark replied:
it needs to increase by the size of the read char each step right?

Watermelon914 replied:
I see what you mean
